Add GlamorousToolkit package

akgrant43 opened the pull request:
###### Description of changes

Add a package for the GlamorousToolkit, a Smalltalk based, moldable development environment.

See <https://gtoolkit.com/>

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,80 @@
+{ stdenv, lib, makeWrapper, fetchzip, pkgs, xorg }:
```
Never use pkgs or xorg

akgrant43 replied:
Hi @AndersonTorres ,
Thanks for your feedback.  I've modified it as you suggested.
Cheers,
Alistair

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+{ stdenv, lib, makeWrapper, fetchzip
+, gtk3
+, glib
+, libX11
+, cairo
+, libglvnd
+, fontconfig
+, freetype
+, libxcb
+, libXrender
+, libXext
+, libuuid
+, dbus
+, libXcursor
+, libXrandr
+, libXi }:
```
```suggestion
{ lib
, stdenv
, fetchzip
, cairo
, dbus
, fontconfig
, freetype
, glib
, gtk3
, libX11
, libXcursor
, libXext
, libXi 
, libXrandr
, libXrender
, libglvnd
, libuuid
, libxcb
, makeWrapper
}:

```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+, libXrandr
+, libXi }:
+stdenv.mkDerivation rec {
+  version = "0.7.2";
+  name = "GlamorousToolkitVM-${version}";
```
```suggestion
  pname = "glamoroustoolkit";
  version = "0.7.2";
```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+  name = "GlamorousToolkitVM-${version}";
+
+  src = fetchzip {
+    url = "https://github.com/feenkcom/gtoolkit-vm/releases/download/v0.7.2/GlamorousToolkit-x86_64-unknown-linux-gnu.zip";
+    sha256 = "Tahe/P0sxmuhSG5JxOkyQGLTOyVrjMEkxvzPc2CBV0Y=";
+    stripRoot = false;
```
```suggestion
    url = "https://github.com/feenkcom/gtoolkit-vm/releases/download/v${finalAttrs.version}/GlamorousToolkit-x86_64-unknown-linux-gnu.zip";
    stripRoot = false;
    hash = "sha256-Tahe/P0sxmuhSG5JxOkyQGLTOyVrjMEkxvzPc2CBV0Y=";
```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+    sha256 = "Tahe/P0sxmuhSG5JxOkyQGLTOyVrjMEkxvzPc2CBV0Y=";
+    stripRoot = false;
+  };
+  sourceRoot = ".";
```
```suggestion

  sourceRoot = ".";
```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+  dontBuild = true;
+  dontPatchELF = true;
+
+  installPhase = ''
```
```suggestion
  installPhase = ''
    runHook preInstall

```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+    chmod +w $out/bin
+    cp ${./GlamorousToolkit-GetImage} $out/bin/GlamorousToolkit-GetImage
+    chmod -w $out/bin
+  '';
```
```suggestion

    runHook postInstall
  '';
```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+    ];
+
+  preFixup = ''
+    echo "libPath=${libPath}"
```
Debugging?

akgrant43 replied:
Yes, removed.

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+    chmod -w $out/bin
+  '';
+
+  libPath = lib.makeLibraryPath [
+      gtk3
+      glib
+      libX11
+      cairo
+      libglvnd
+      fontconfig
+      freetype
+      libxcb
+      libXrender
+      libXext
+      libuuid
+      dbus
+      libXcursor
+      libXrandr
+      libXi
+      stdenv.cc.cc.lib
+    ];
+
+  preFixup = ''
```
```suggestion
preFixup = let
    libPath = lib.makeLibraryPath [
      cairo
      dbus
      fontconfig
      freetype
      glib
      gtk3
      libX11
      libXcursor
      libXext
      libXi
      libXrandr
      libXrender
      libglvnd
      libuuid
      libxcb
      stdenv.cc.cc.lib
    ];
  in ''
```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+, libXcursor
+, libXrandr
+, libXi }:
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+  dontPatchELF = true;
+
+  installPhase = ''
+    mkdir $out
+    cp -a $src/bin $src/lib $out/
+    chmod +w $out/bin
+    cp ${./GlamorousToolkit-GetImage} $out/bin/GlamorousToolkit-GetImage
+    chmod -w $out/bin
```
This is strange. Have you tried to create the directory tree first?

```suggestion
    install -d $out/bin $out/lib
    cp -r $src/bin $src/lib $out/
    cp ${./GlamorousToolkit-GetImage} $out/bin/GlamorousToolkit-GetImage
```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,95 @@
+    ln -s $out/lib/libgit2.so $out/lib/libgit2.so.1.1
+  '';
+
+  meta = with lib; {
+    homepage = "https://gtoolkit.com";
+    description = "The GlamorousToolkit Development Environment";
+    license = licenses.mit;
+    platforms = platforms.linux;
+    maintainers = [ lib.maintainers.akgrant43 ];
+  };
```
```suggestion
  meta = {
    homepage = "https://gtoolkit.com";
    description = "The GlamorousToolkit Development Environment";
    license = lib.licenses.mit;
    maintainers = [ lib.maintainers.akgrant43 ];
    platforms = lib.platforms.linux;
  };
```

AndersonTorres commented:
Are you editing locally?

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:
```diff
@@ -0,0 +1,23 @@
+	exit 1
+fi
+
+wget -q --show-progress https://github.com/feenkcom/gtoolkit/releases/download/${GTVERSION}/GlamorousToolkit-Linux-x86_64-${GTVERSION}.zip -O gt-image.zip
```
```suggestion
wget -q --show-progress \
  -O gt-image.zip \
  https://github.com/feenkcom/gtoolkit/releases/download/${GTVERSION}/GlamorousToolkit-Linux-x86_64-${GTVERSION}.zip 
```

akgrant43 commented:
> Are you editing locally?

Yes, I'm not at all familiar with github review tools, and most of the changes make sense, so I've just accepted all the changes and am testing locally (although my time is very sporadic today).

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,102 @@
+
+  src = fetchzip {
+    url = "https://github.com/feenkcom/gtoolkit-vm/releases/download/${finalAttrs.version}/GlamorousToolkit-x86_64-unknown-linux-gnu.zip";
+    sha256 = "Tahe/P0sxmuhSG5JxOkyQGLTOyVrjMEkxvzPc2CBV0Y=";
+    stripRoot = false;
```
```suggestion
    stripRoot = false;
    hash = "sha256-Tahe/P0sxmuhSG5JxOkyQGLTOyVrjMEkxvzPc2CBV0Y=";
```

AndersonTorres reviewed (changes requested):
Squash the commits into one, except the one updating maintainers list

akgrant43 commented:
Hi @AndersonTorres ,

Thanks again for all your feedback.

> Squash the commits into one, except the one updating maintainers list

I've squashed all the commits from today in to a single commit, however there are merges from master prior to that and the rebase command tries to load over 4000 commits and fails.

Would you prefer I close this PR and start again?

Thanks,
Alistair

AndersonTorres commented:
Have you tried to rebase onto upstream/master, instead of merge?

akgrant43 commented:
I did an interactive rebase on the branch from which the PR is sourced: https://github.com/akgrant43/nixpkgs/tree/glamoroustoolkit-initial, e.g.

```
$ git status
On branch glamoroustoolkit-initial
Your branch is up to date with 'akgrant43/glamoroustoolkit-initial'.
...

$ git rebase -i HEAD~14
```

So the commit shows a single entry for today: https://github.com/akgrant43/nixpkgs/commits/glamoroustoolkit-initial

I think the issue is that I merged master back in to `glamoroustoolkit-initial` between the initial set of commits from 3 weeks ago and today's work.

akgrant43 commented:
OK, I think I understand what you mean now.  I've done a `git rebase master glamoroustoolkit-initial` which seems to have corrected the order of the commits.

I'll try squashing the remaining commits next (excluding the maintainers list).

akgrant43 commented:
OK, now it is just two commits.  I hope I've got it correct. :-)

akgrant43 commented:
I missed one of your suggestions earlier (sha256 format), it's included now.

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,102 @@
+    license = lib.licenses.mit;
+    maintainers = [ lib.maintainers.akgrant43 ];
+    platforms = lib.platforms.linux;
+  };
```
```suggestion
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
  };
```

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,102 @@
+    description = "The GlamorousToolkit Development Environment";
+    license = lib.licenses.mit;
+    maintainers = [ lib.maintainers.akgrant43 ];
+    platforms = lib.platforms.linux;
```
```suggestion
    platforms = "x86_64-linux";
```

akgrant43 commented:
I've integrated and squashed the latest two changes.

akgrant43 commented:
It looks like the platforms value should be a list, instead of a string, i.e.:

```
    platforms = [ "x86_64-linux" ];
```

You've just approved this, I believe I should commit this change, but please let me know what you think.

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/default.nix:
```diff
@@ -0,0 +1,103 @@
+  version = "0.7.2";
+
+  src = fetchzip {
+    url = "https://github.com/feenkcom/gtoolkit-vm/releases/download/${finalAttrs.version}/GlamorousToolkit-x86_64-unknown-linux-gnu.zip";
```
```suggestion
    url = "https://github.com/feenkcom/gtoolkit-vm/releases/download/v${finalAttrs.version}/GlamorousToolkit-x86_64-unknown-linux-gnu.zip";
```

akgrant43 replied:
Good pickup, thanks!

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:
```diff
@@ -0,0 +1,25 @@
+	read -p "Continue? (Y/N): " confirm && [[ $confirm == [yY] || $confirm == [yY][eE][sS] ]] || exit 1
+fi
+
+GTVERSION=$(curl -s https://api.github.com/repos/feenkcom/gtoolkit/releases/latest 2>/dev/null | grep tag_name | cut -d '"' -f 4)
```
Is it OK to use a version that not matches the installed version?

akgrant43 replied:
You mean that the VM version doesn't match the image version?

GT is built on smalltalk, typically the VM changes multiple times a year while the image changes multiple times per day.

AndersonTorres replied:
Nice!

akgrant43 replied:
Thanks. :-)

The summary at the bottom of the page (GitHub conversation page) is saying that you are requesting a change here:

![2023-08-06_22-39](https://github.com/NixOS/nixpkgs/assets/2062166/170383a5-b90f-486c-933a-5db10b675144)

However the conversation is showing as resolved, and there's nothing to commit, so I'm not sure what's required.

Thanks again for all your help.

AndersonTorres replied:
GitHub requires a final review in order to remove these.

AndersonTorres commented:
Also, fix the problems.

AndersonTorres commented:
> It looks like the platforms value should be a list, instead of a string, i.e.:
> 
> ```
>     platforms = [ "x86_64-linux" ];
> ```

That's right. I haven't noticed.

> 
> You've just approved this, I believe I should commit this change, but please let me know what you think.

akgrant43 commented:
The `platforms` issue has been resolved.  Unless I've missed something, I'm not aware of any other issues at the moment.  I built the package in a clean VM to check and it's fine.

AndersonTorres reviewed (changes requested):
Finally, squash the two latest commits into only one.

akgrant43 commented:
> Finally, squash the two latest commits into only one.

Done.

AndersonTorres commented on pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:
```diff
@@ -0,0 +1,25 @@
+	exit 1
+fi
+
+wget -q --show-progress \
```
Well, you are already using curl to download page info and discover what image to download next.
Why not to use it again here instead of wget?

akgrant43 replied:
No reason :-)
Changed, squashed and pushed.

AndersonTorres reviewed (approved):
Let's wait one day to merge.

akgrant43 commented:
Thanks!

akgrant43 commented:
> Let's wait one day to merge.

I was on holiday for 3 weeks and then travelling for work.  I think one day has passed. :-)

Is there anything else I need to do for this to be merged?

Thanks!

AndersonTorres commented:
Fix conflicts

akgrant43 committed:
maintainers-list: add akgrant43

akgrant43 committed:
glamoroustoolkit: init at v0.7.2

Includes significant feedback from Anderson Torres to refactor
derivation to meet guidelines.

Co-authored-by: Anderson Torres <torres.anderson.85@protonmail.com>

akgrant43 commented:
Hi @AndersonTorres,

The merge conflict is resolved.  Please let me know if there's anything else I should do.

Thanks!

akgrant43 committed:
glamoroustoolkit: 0.7.2 -> 0.7.3

akgrant43 commented:
Thanks again for all your help, @AndersonTorres !
